File: src/assay/minimum.py
"""First-occurrence minimum composition over normalized components."""

from __future__ import annotations

from assay.composite import inputs_hash, interval_or_none
from assay.contracts import (
    Component,
    ExplainedComponent,
    Interval,
    Method,
    MinimumRequest,
    Operation,
    ScoreResult,
)
from assay.normalize import normalize


def _normalized(component: Component, request: MinimumRequest, value: float) -> float:
    return normalize(value, component.scale, request.clamp)


def _explain(component: Component, request: MinimumRequest) -> ExplainedComponent:
    candidate = _normalized(component, request, component.value)
    return ExplainedComponent(
        id=component.id,
        raw=component.value,
        normalized=candidate,
        operation=Operation.ADD,
        coefficient=1.0,
        contribution=candidate,
    )


def _bounds(component: Component, request: MinimumRequest) -> tuple[float, float]:
    interval = component.interval
    if interval is None:
        point = _normalized(component, request, component.value)
        return point, point
    first = _normalized(component, request, interval.low)
    second = _normalized(component, request, interval.high)
    return min(first, second), max(first, second)


def _propagated_bounds(request: MinimumRequest) -> tuple[float, float]:
    bounds = tuple(_bounds(component, request) for component in request.components)
    return min(low for low, _ in bounds), min(high for _, high in bounds)


def _result_interval(request: MinimumRequest) -> Interval | None:
    if not any(component.interval is not None for component in request.components):
        return None
    return interval_or_none(*_propagated_bounds(request))


def minimum(request: MinimumRequest) -> ScoreResult:
    """Return the first lowest normalized candidate and identify it explicitly."""
    validated = MinimumRequest.model_validate(request)
    rows = tuple(_explain(component, validated) for component in validated.components)
    selected = min(rows, key=lambda row: row.contribution)
    return ScoreResult(
        method=Method(id=validated.method, version=validated.method_version),
        score=selected.contribution,
        interval=_result_interval(validated),
        components=rows,
        inputs_hash=inputs_hash(validated),
        selected_component_id=selected.id,
    )

File: ts/src/minimum.ts
import {
  type Component,
  type ExplainedComponent,
  type Interval,
  type MinimumRequest,
  parseRequest,
  parseScoreResult,
  type ScoreResult,
} from "./contracts.js";
import { ContractCode, ContractValidationError } from "./errors.js";
import { normalize } from "./normalize.js";
import { inputsHash } from "./requestHash.js";

function bounds(
  component: Component,
  request: MinimumRequest,
): readonly [number, number] {
  if (component.interval === null) {
    const point = normalize(component.value, component.scale, request.clamp);
    return [point, point];
  }
  const first = normalize(
    component.interval.low,
    component.scale,
    request.clamp,
  );
  const second = normalize(
    component.interval.high,
    component.scale,
    request.clamp,
  );
  return [Math.min(first, second), Math.max(first, second)];
}

function candidateInterval(
  component: Component,
  request: MinimumRequest,
): Interval | null {
  if (component.interval === null) return null;
  const [low, high] = bounds(component, request);
  return low === high ? null : { low, high };
}

function row(
  component: Component,
  request: MinimumRequest,
): ExplainedComponent {
  const normalized = normalize(component.value, component.scale, request.clamp);
  return {
    id: component.id,
    raw: component.value,
    normalized,
    declared_weight: null,
    operation: "add",
    coefficient: 1,
    contribution: normalized,
    contribution_interval: candidateInterval(component, request),
  };
}

function resultInterval(request: MinimumRequest): Interval | null {
  if (!request.components.some((component) => component.interval !== null))
    return null;
  let low = Number.POSITIVE_INFINITY;
  let high = Number.POSITIVE_INFINITY;
  for (const component of request.components) {
    const [candidateLow, candidateHigh] = bounds(component, request);
    low = Math.min(low, candidateLow);
    high = Math.min(high, candidateHigh);
  }
  return low === high ? null : { low, high };
}

function firstMinimum(
  rows: ReadonlyArray<ExplainedComponent>,
): ExplainedComponent {
  const first = rows[0] as ExplainedComponent;
  let selected = first;
  for (const candidate of rows.slice(1)) {
    if (candidate.contribution < selected.contribution) selected = candidate;
  }
  return selected;
}

export function minimum(input: MinimumRequest): ScoreResult {
  const request = parseRequest(input);
  if (request.method !== "minimum") {
    throw new ContractValidationError(ContractCode.INVALID_METHOD);
  }
  const rows = request.components.map((component) => row(component, request));
  const selected = firstMinimum(rows);
  return parseScoreResult({
    schema: "assay.result/v1",
    method: { id: request.method, version: request.method_version },
    score: selected.contribution,
    interval: resultInterval(request),
    clamp: request.clamp,
    intercept: null,
    weight_total: null,
    components: rows,
    inputs_hash: inputsHash(request),
    selected_component_id: selected.id,
  });
}
